### PR TITLE
Add bucketization for usage

### PR DESCRIPTION
**Pull request by yojs**
*Fixes #:* #449

*Description of changes:* Adds a way to create bucket calculators based on rca.conf settings. So say we want to add configurable bucket sizes for old-gen, so you add a new object to `rca.conf` (all three), something like this:
```
{
...
"bucketization": {
    "old-gen": {
      "UNDER_UTILIZED": ..,
      "HEALTHY_WITH_BUFFER": ..,
      "HEALTHY": ..
    }
}
...
```

The bucketCalculator hangs off the `RcaConf` and `Deciders` have access to `RcaConf` object. So, in order to bucketize the value, all a decider has to do is

`rcaConf.getBucketizationSettings("old-gen").compute(<current-value>)`

*Tests:* Added unit tests.

*If new tests are added, how long do the new ones take to complete* 
```
  Test testCreateBucketAssigner PASSED (3.3s)
  Test testMisSpelledTunableName PASSED
  Test testNoThresholds PASSED
  Test testNoBucketizationSettings PASSED

SUCCESS: Executed 4 tests in 6.9s
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by yojs** [comment]
```
Test nodeRoleChange FAILED (2s)

 429 tests completed, 1 failed
FAILURE: Executed 429 tests in 13m 41s (1 failed)
```

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/bucket/neo/BucketAssignerTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    RcaConf rcaConf = new RcaConf(Paths.get(RcaConsts.TEST_CONFIG_PATH, "rca.conf").toString());
+    rcaConf.getBucketizationSettings("base1");
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testNoThresholds() {
+    RcaConf rcaConf = new RcaConf(Paths.get(RcaConsts.TEST_CONFIG_PATH, "bucketization", "bucketize_no_thresholds.json").toString());
+    rcaConf.getBucketizationSettings("base1");
+  }
+}
```
Comment (sruti1312) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/bucket/neo/BucketAssignerTest.java:49]: nit: new line

File: src/test/resources/rca/bucketization/bucketize.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      "HEALTHY_WITH_BUFFER": 40.0,
+      "HEALTHY": 75.0
+    },
+    "base3": {
+      "UNDER_UTILIZED": 30.1,
+      "HEALTHY_WITH_BUFFER": 40.23456,
+      "HEALTHY": 75.0
+    }
+  }
+}
```
Comment (sruti1312) [inline src/test/resources/rca/bucketization/bucketize.json:19]: nit: new line

File: src/test/resources/rca/bucketization/bucketize_no_thresholds.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+  "bucketization": {
+  }
+}
```
Comment (sruti1312) [inline src/test/resources/rca/bucketization/bucketize_no_thresholds.json:4]: nit: new line

Comment (sruti1312) [inline src/test/resources/rca/bucketization/bucketize_no_thresholds.json:4]: Can we use a string (in the same file) and pass it to rca.conf 

Ref:
https://github.com/opendistro-for-elasticsearch/performance-analyzer-rca/blob/master/src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyCacheMaxSizeActionTest.java#L113

File: src/test/resources/rca/bucketization/bucketize.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (sruti1312) [inline src/test/resources/rca/bucketization/bucketize.json:1]: Same as below. Can we make this a string and pass to rca.conf. 

It makes the mapping better (which test is using which input)

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/bucket/neo/BucketAssignerTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.bucket.BucketCalculator;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.bucket.UsageBucket;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.core.RcaConf;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.util.RcaConsts;
+import java.nio.file.Paths;
+import java.util.Map;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class BucketAssignerTest {
```
Comment (sruti1312) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/bucket/neo/BucketAssignerTest.java:12]: What are you thoughts on adding these to RcaConfTest.java

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/core/RcaConf.java [hunk 6]
```diff
@@ -306,4 +313,69 @@ private boolean updateRcaConf(String originalFilePath, final Set<String> mutedRc
...
+  private void constructTunableResourceToUsageBucket() {
+    Map<String, Object> tunableSettingsMap = conf.getBucketizationTunings();
+    if (tunableSettingsMap == null) {
+      // The bucketization key does not exist in the rca.conf. So we cannot construct BucketCalculators.
+      return;
+    }
+    for (Map.Entry<String, Object> entry : tunableSettingsMap.entrySet()) {
+      String currentTunable = entry.getKey();
+      if (entry.getValue() instanceof Map) {
+        Map<UsageBucket, Double> usageBucketLimitMap = new HashMap<>();
```
Comment (sruti1312) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/core/RcaConf.java:361]: Can we use a ImmutableMap here?

**Review by sruti1312 — CHANGES_REQUESTED** [review]

**Review by vigyasharma — COMMENTED** [review]
No new comments from me, will wait for you to address @sruti1312's comments before approving.

File: src/test/resources/rca/bucketization/bucketize.json (hunk 4 above)
Comment (yojs) [reply src/test/resources/rca/bucketization/bucketize.json]: got it !

**Review by sruti1312 — APPROVED** [review]

**Review by vigyasharma — APPROVED** [review]

**Pull request merged**